src/sentry/auth/providers/saml2.py

```diff
@@ -391,19 +390,27 @@ def dispatch(self, request, organization_slug):
         if provider is None:
             raise Http404
 
-        if 'options' not in provider.config or not provider.config['options'].get('options_slo', False):
-            raise Exception("SLO is disabled, can't process that action")
-
         saml_config = provider.build_saml_config(organization_slug)
-
         auth = provider.build_auth(request, saml_config)
-        dscb = lambda: logout(request)
-        next = auth.process_slo(delete_session_cb=dscb)
-        request.user = AnonymousUser()
-        if not next:
-            next = get_login_url()
 
-        return self.redirect(next)
+        # If SLS is disabled do not log them out, but continue the SLS logout flow
```

Comment:
If Single Logout is disabled and Sentry replies back to the IdP, that means that all was ok (system closed the local session and redirecting back) so the user will expect that he was logged out of all the apps connected with the IdP.

If SLS is disabled, IdP shouldn't have registered the Sentry SLS endpoint so inform of the fact that SLS is disabled can warn the administrator to take care of that issue, by removing that endpoint from the IdP.

That is the behavior you can see in most of the SAML integrations that supports Single Logout.

Reply:
Is there a SAML response we can make that says 'Hey we didn't log out the user, but keep going with the current logout flow'

Because right now when logging out of OneLogin from their website, it redirects to sentry and will just error out when SLO is disabled, leaving the user with an annoying message.

Bonus question: Does SAML support a 'forced logout' type message. Currently, it's the user who has to initiate the logout. Is there a way for the administrator to logout a user?

Reply:
There is a "we didn't log out the user" reply but not all Identity Providers support it, so maybe you will end in an error page on the IdP side.

When user log out from logout, it tries to logout at sentry because Sentry's SLS endpoint is registered at OneLogin. If Logout is not supported at Sentry, SLS endpoint should be removed from Onelogin, so it never gonna try log out the user at Sentry... so think in the scenario of Sentry's SLS registered at Onelogin but Logout disabled at Sentry as a configuration error made by the admin.

In SAML most of the IdPs implements Logout using front-channel, and not the back-channel, so browser cookies of the user are involved on the logout process, so an Admin can't "force a massive logout" to initiate a SAML SLO flow of the users, the user may initiate that process, so maybe enabling some flag on system, so when the user access, is forced to initiate the SLO process.

SLO is complex, if you want to read more about it, see:
- https://wiki.shibboleth.net/confluence/display/CONCEPT/SLOIssues
- https://pdfs.semanticscholar.org/e4b0/20decd4e522f1390a89b73f6449f6766e0b1.pdf

Reply:
@mattrobenolt @dcramer: What do you guys think about this? If there's a mismatch between SLOs being enabled on sentry's side and on the onelogin/IdP side, then what should the user experience be:

1. Ideally we tell the service provider we didn't log them out. Presumably that's done by by setting the `StatusCode` field in the [SLO response](https://www.samltool.com/generic_slo_res.php) to some failure constant yes?

   Though according to what @pitbulk is saying here, some Identity providers will then just error out on the users side

2. Do we want to error out on our side. I think this is really confusing for the user, since the SLO flow *starts* from the IdP (onelogin) side, redirects to sentry, then redirects *back* to onelogin. However, if we raise an error on our end, then the user just lands on on their sentry dashboard with a message telling them they weren't logged out (super confusing).

3. Do we just always tell the IdP that we've logged out the user.. even if we haven't. @pitbulk's point stands then

   > If Single Logout is disabled and Sentry replies back to the IdP, that means that all was ok (system closed the local session and redirecting back) so the user will expect that he was logged out of all the apps connected with the IdP.

4. Don't make turning SLO on or off an option. It's always on